test/framework/netns/netns.go

```diff
@@ -0,0 +1,270 @@
+package netns
+
+import (
+	"fmt"
+	"net"
+	"runtime"
+	"strings"
+	"syscall"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+)
+
+// CLONE_NEWNET requires Linux Kernel 3.0+
+
+type Veth struct {
+	veth        *netlink.Veth
+	name        string
+	peerName    string
+	address     string
+	cidr        string
+	peerAddress string
+	peerCIDR    string
+}
+
+func (v *Veth) Veth() *netlink.Veth {
+	return v.veth
+}
+
+func (v *Veth) PeerName() string {
+	return v.peerName
+}
+
+func (v *Veth) Address() string {
+	return v.address
+}
+
+func (v *Veth) Cidr() string {
+	return v.cidr
+}
+
+func (v *Veth) PeerAddress() string {
+	return v.peerAddress
+}
+
+func (v *Veth) PeerCIDR() string {
+	return v.peerCIDR
+}
+
+func (v *Veth) Name() string {
+	return v.name
+}
+
+type NetNS struct {
+	name       string
+	ns         netns.NsHandle
+	originalNS netns.NsHandle
+	veth       *Veth
+}
+
+func (ns *NetNS) Name() string {
+	return ns.name
+}
+
+func (ns *NetNS) Veth() *Veth {
+	return ns.veth
+}
+
+func (ns *NetNS) Set() error {
+	if err := netns.Set(ns.ns); err != nil {
+		return fmt.Errorf("cannot switch to the network namespace %q: %s", ns.ns.String(), err)
+	}
+
+	return nil
+}
+
+func (ns *NetNS) Exec(callback func()) <-chan error {
+	done := make(chan error)
+
+	go func() {
+		defer ginkgo.GinkgoRecover()
+		defer close(done)
+
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+
+		if err := ns.Set(); err != nil {
+			done <- fmt.Errorf("cannot set the namespace %q: %s", ns.name, err)
+		}
+		defer ns.Unset()
+
+		callback()
```

Comment:
yes, it spawns up the goroutine to isolate the thread, I think the only problem could be if we would switch namespaces inside the callback, but as we are the owners of the code we should be able to catch this during the reviews

Reply:
It's not switching namespaces in the callback that is the problem. It's spawning a goroutine. If the callback spawns a goroutine, it will most likely be in the incorrect namespace. The "Gotchas" section in the linked blog post shows this.

Reading a little further the docs on LockOSThread, it make sense. When you call LockOSThread, the current goroutine will have exclusive access to that thread. Any goroutine spawned will then be scheduled on a _different_ OS thread, which will be in the wrong namespace.  The blog post refers to go 1.10, but I tested the example on 1.17 and got the same results.

I really think we need a big warning on this that any callbacks must not spawn any goroutines that depend on the net ns, lest they end up in the wrong one.

Reply:
addressed by changing name from `Exec` to `UnsafeExec` and added appropriate comment